In [ ]:
# Packages
import pandas as pd

In [ ]:
# Importing data
df = pd.read_csv("HousingPrice.csv", sep = ";")
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [ ]:
# Checking datatypes
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
                  ...   
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
SalePrice          int64
Length: 81, dtype: object

In [ ]:
# Extracting the features that we want
df1 = df[["MSSubClass", "MSZoning", "LotFrontage", "LotArea", "LotShape", "LandContour", "Utilities",
          "PoolArea", "YrSold", "SaleCondition", "SalePrice"]]
df1.head()

MSSubClass MSZoning  LotFrontage  LotArea LotShape LandContour Utilities  \
0          60       RL         65.0     8450      Reg         Lvl    AllPub   
1          20       RL         80.0     9600      Reg         Lvl    AllPub   
2          60       RL         68.0    11250      IR1         Lvl    AllPub   
3          70       RL         60.0     9550      IR1         Lvl    AllPub   
4          60       RL         84.0    14260      IR1         Lvl    AllPub   

   PoolArea  YrSold SaleCondition  SalePrice  
0         0    2008        Normal     208500  
1         0    2007        Normal     181500  
2         0    2008        Normal     223500  
3         0    2006       Abnorml     140000  
4         0    2008        Normal     250000

In [ ]:
# Encoding some features
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df1["MSZoning"] = le.fit_transform(df1["MSZoning"])
df1["LotShape"] = le.fit_transform(df1["LotShape"])
df1["LandContour"] = le.fit_transform(df1["LandContour"])
df1["Utilities"] = le.fit_transform(df1["Utilities"])
df1["SaleCondition"] = le.fit_transform(df1["SaleCondition"])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pan

In [ ]:
# Converting the relevant features to categorical type, and the "SalePrice" is a float
df1["MSSubClass"] = df1["MSSubClass"].astype("category")
df1["MSZoning"] = df1["MSZoning"].astype("category")
df1["LotShape"] = df1["LotShape"].astype("category")
df1["LandContour"] = df1["LandContour"].astype("category")
df1["Utilities"] = df1["Utilities"].astype("category")
df1["YrSold"] = df1["YrSold"].astype("category")
df1["SaleCondition"] = df1["SaleCondition"].astype("category")
df1["SalePrice"] = df1["SalePrice"].astype("float64")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

In [ ]:
# Checking datatypes to ensure proper encoding
df1.dtypes

MSSubClass       category
MSZoning         category
LotFrontage       float64
LotArea             int64
LotShape         category
LandContour      category
Utilities        category
PoolArea            int64
YrSold           category
SaleCondition    category
SalePrice         float64
dtype: object

In [ ]:
# Checking for null values
df1.isnull().sum()

MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
LotShape           0
LandContour        0
Utilities          0
PoolArea           0
YrSold             0
SaleCondition      0
SalePrice          0
dtype: int64

We observe **259** null values from the variable "LotFrontage", we will simply ignore this variable in our subsequent
analysis, as since we have less data, dropping it will make us lose a lot of the data

In [ ]:
# Having a broad overview of our data
df1

MSSubClass MSZoning  LotFrontage  LotArea LotShape LandContour Utilities  \
0            60        3         65.0     8450        3           3         0   
1            20        3         80.0     9600        3           3         0   
2            60        3         68.0    11250        0           3         0   
3            70        3         60.0     9550        0           3         0   
4            60        3         84.0    14260        0           3         0   
...         ...      ...          ...      ...      ...         ...       ...   
1455         60        3         62.0     7917        3           3         0   
1456         20        3         85.0    13175        3           3         0   
1457         70        3         66.0     9042        3           3         0   
1458         20        3         68.0     9717        3           3         0   
1459         20        3         75.0     9937        3           3         0   

      PoolArea YrSold SaleCondition  SalePrice  
0            0   2008             4   208500.0  
1            0   2007             4   181500.0  
2            0   2008             4   223500.0  
3            0   2006             0   140000.0  
4            0   2008             4   250000.0  
...        ...    ...           ...        ...  
1455         0   2007             4   175000.0  
1456         0   2010             4   210000.0  
1457         0   2010             4   266500.0  
1458         0   2010             4   142125.0  
1459         0   2008             4   147500.0  

[1460 rows x 11 columns]

In [ ]:
# Splitting the data into train and test sets
from sklearn.model_selection import train_test_split

X = df1[["MSSubClass", "MSZoning", "LotArea", "LotShape", "LandContour", "Utilities", "PoolArea", 
         "YrSold", "SaleCondition"]]
y = df1[["SalePrice"]]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.4)

In [ ]:
# Building the model
from sklearn import linear_model

# Lasso (L1 regularisation) regression
lasso_reg = linear_model.Lasso(alpha = 50, max_iter = 1000, tol = 0.1)
lasso_reg.fit(X_train, y_train)
lasso_reg.score(X_test, y_test)

0.21200578170468554

In [ ]:
# Ridge (L2 regularisation) regression
ridge_reg = linear_model.Ridge(alpha = 50, max_iter = 1000, tol = 0.1)
ridge_reg.fit(X_train, y_train)
ridge_reg.score(X_test, y_test)

0.20841830994131383

Clearly, these low scores are indicative of real-world data versus working with fictitious data. Also, we can improve the score by adding more observations!